src/rtree.c

```diff
@@ -104,90 +150,137 @@ rtree_node_init(tsdn_t *tsdn, rtree_t *rtree, unsigned level,
-		child = (rtree_elm_t *)atomic_load_p(&elm->child_or_extent,
+		node = (rtree_node_elm_t *)atomic_load_p(&elm->child,
 		    ATOMIC_RELAXED);
 	} else {
-		child = (rtree_elm_t *)atomic_load_p(&elm->child_or_extent,
+		node = (rtree_node_elm_t *)atomic_load_p(&elm->child,
 		    ATOMIC_ACQUIRE);
 	}
 
-	assert(!dependent || child != NULL);
-	return child;
+	assert(!dependent || node != NULL);
+	return node;
 }
 
-static rtree_elm_t *
-rtree_child_read(tsdn_t *tsdn, rtree_t *rtree, rtree_elm_t *elm, unsigned level,
-    bool dependent) {
-	rtree_elm_t *child;
+static rtree_node_elm_t *
+rtree_child_node_read(tsdn_t *tsdn, rtree_t *rtree, rtree_node_elm_t *elm,
+    unsigned level, bool dependent) {
+	rtree_node_elm_t *node;
 
-	child = rtree_child_tryread(elm, dependent);
-	if (!dependent && unlikely(!rtree_node_valid(child))) {
-		child = rtree_node_init(tsdn, rtree, level + 1,
-		    &elm->child_or_extent);
+	node = rtree_child_node_tryread(elm, dependent);
+	if (!dependent && unlikely(!rtree_node_valid(node))) {
+		node = rtree_node_init(tsdn, rtree, level + 1, &elm->child);
 	}
-	assert(!dependent || child != NULL);
-	return child;
+	assert(!dependent || node != NULL);
+	return node;
 }
 
-static rtree_elm_t *
-rtree_subtree_tryread(rtree_t *rtree, bool dependent) {
-	rtree_elm_t *subtree;
+static rtree_leaf_elm_t *
+rtree_child_leaf_tryread(rtree_node_elm_t *elm, bool dependent) {
+	rtree_leaf_elm_t *leaf;
+
 	if (dependent) {
-		subtree = (rtree_elm_t *)atomic_load_p(&rtree->root,
+		leaf = (rtree_leaf_elm_t *)atomic_load_p(&elm->child,
 		    ATOMIC_RELAXED);
 	} else {
-		subtree = (rtree_elm_t *)atomic_load_p(&rtree->root,
+		leaf = (rtree_leaf_elm_t *)atomic_load_p(&elm->child,
 		    ATOMIC_ACQUIRE);
 	}
-	assert(!dependent || subtree != NULL);
-	return subtree;
+
+	assert(!dependent || leaf != NULL);
+	return leaf;
 }
 
-static rtree_elm_t *
-rtree_subtree_read(tsdn_t *tsdn, rtree_t *rtree, bool dependent) {
-	rtree_elm_t *subtree = rtree_subtree_tryread(rtree, dependent);
-	if (!dependent && unlikely(!rtree_node_valid(subtree))) {
-		subtree = rtree_node_init(tsdn, rtree, 0, &rtree->root);
+static rtree_leaf_elm_t *
+rtree_child_leaf_read(tsdn_t *tsdn, rtree_t *rtree, rtree_node_elm_t *elm,
+    unsigned level, bool dependent) {
+	rtree_leaf_elm_t *leaf;
+
+	leaf = rtree_child_leaf_tryread(elm, dependent);
+	if (!dependent && unlikely(!rtree_leaf_valid(leaf))) {
+		leaf = rtree_leaf_init(tsdn, rtree, &elm->child);
 	}
-	assert(!dependent || subtree != NULL);
-	return subtree;
+	assert(!dependent || leaf != NULL);
+	return leaf;
 }
 
-rtree_elm_t *
-rtree_elm_lookup_hard(tsdn_t *tsdn, rtree_t *rtree, rtree_ctx_t *rtree_ctx,
+rtree_leaf_elm_t *
+rtree_leaf_elm_lookup_hard(tsdn_t *tsdn, rtree_t *rtree, rtree_ctx_t *rtree_ctx,
     uintptr_t key, bool dependent, bool init_missing) {
-	rtree_elm_t *node = init_missing ? rtree_subtree_read(tsdn, rtree,
-	    dependent) : rtree_subtree_tryread(rtree, dependent);
+	rtree_node_elm_t *node;
```

Comment:
I considered this, but the compiler doesn't complain about the unused variable, apparently because there is still dead code that references it.